### DISCUSSION TITLE
Is there a grammar for PML?

### DISCUSSION THREAD

**Discussion by yamahito** [pml-lang/pml-userman#13] [Q&A] [answered]
Hi, 

Is there an EBNF or similar grammar for PML?

**Comment by pml-lang** [comment, marked as answer]
Hello Tomos,

You can find a grammar definition [here](https://www.pml-lang.dev/docs/syntax/ebnf.html). Note that this is a draft, not public yet (i.e. not accessible from the menu on PML's website).

PML uses the [PDML](https://pdml-lang.github.io/) syntax. The following documents describe  _basic_ PDML (without optional extensions like attributes):
- [Specification](https://pdml-lang.github.io/docs/basic-specification/index.html)
- [Grammar (EBNF-like)](https://pdml-lang.github.io/docs/ebnf/index.html)
- [Railroad Diagrams](https://pdml-lang.github.io/docs/railroad/index.html)

**Reply by yamahito** [reply]
Thanks, Christian.

I am an implementor/committee member for the community group for Invisible XML: https://invisiblexml.org - it looks like it should be fairly easy to make the XML technology stack (xslt, xquery etc) available for PML documents, so it might make an interesting use case (if I ever get my parser working performantly!)

I mean this as an observation, not a criticism, but isn't PML essentially just a different serialization/representation of an XML application?

**Reply by pml-lang** [reply]
> Invisible XML

Very interesting project! The ability to convert _any_ structured data into XML by just providing a grammar is very powerful.

> it looks like it should be fairly easy to make the XML technology stack (xslt, xquery etc) available for PML documents

Yes. A customized PML to XML conversion is already implemented in PDML's parser. The _outdated_ article [Open-Source Parser For Practical XML (pXML)](https://pdml-lang.github.io/docs/open-source-parser-for-practical-xml/index.html) shows how it works.

I guess the conversion could also be achieved by using Invisible XML. That might indeed be an interesting use case.

> isn't PML essentially just a different serialization/representation of an XML application?

Yes, at its most basic level, PDML is 'just another text format' to represent data. The rationale for a new format is explained in chapter [PDML vs XML/JSON/YAML](https://pdml-lang.github.io/docs/introduction/index.html#ch__9).
